app/Utils/Traits/GeneratesCounter.php

```diff
@@ -420,8 +420,29 @@ private function checkEntityNumber($class, $entity, $counter, $padding, $pattern
         $check = false;
         $check_counter = 1;
 
+        /*
+         * Optimization for getting the next number without iterating the first 100 items & after that returning random number (if numbers already exists)
+         *
+         * See also:
+         * https://github.com/invoiceninja/invoiceninja/pull/8231
+         *
+        $number_placeholder = '{number}';
+        $latest = $class::where('company_id', $entity->company_id)->where('number', '<>', '', 'and')->orderBy('created_at', 'desc')->first();
```

Comment:
This is a problem as it does not guarantee that the actual number is the latest. This is especially true if a user if using different counters per client / group

Reply:
Hey @turbo124 I commented out this whole section, so it's not active, just archived for may later use. But I can also remove this whole comment from merge request for "cleaner" code.

Btw line 438 is checking, if there's a latest counter & $latest_counter > $counter. Only then counter gets replaced with latest.

But as already said, if you don't want to have this code as comment, I'll remove it. I don't really need it, its just a small optimization to avoid the 100 while loops & after that generating random number (only happens e.g. if you reset counters by mistake & get numbers which are already in use).

PS: I can also create a new MR for that (then uncommented), if you want to discuss this in separate topic :-)